docs/development/module/off-chain-workers.md

```diff
@@ -0,0 +1,383 @@
+      system::CheckNonce::<Runtime>::from(index),
+      system::CheckWeight::<Runtime>::new(),
+      transaction_payment::ChargeTransactionPayment::<Runtime>::from(tip),
+    );
+    let raw_payload = SignedPayload::new(call, extra).ok()?;
+    let signature = TSigner::sign(public, &raw_payload)?;
+    let address = Indices::unlookup(account);
+    let (call, extra, _) = raw_payload.deconstruct();
+    Some((call, (address, signature, extra)))
+  }
+}
+```
+
+Inside the `contruct_runtime!` macro where you put all various packages as your runtime enum, add an additional parameter `ValidateUnsigned` if you are using unsigned transactions in off-chain workers.
+
+**runtime/src/lib.rs**
+
+```rust
+construct_runtime!(
+  pub enum Runtime where
+    Block = Block,
+    NodeBlock = opaque::Block,
+    UncheckedExtrinsic = UncheckedExtrinsic
+  {
+    // ...
+    // ...
+    // For using unsigned transactions
+    MyRuntime: my_runtime::{ Module, Call, Storage, Event<T>, transaction_validity::ValidateUnsigned }
+
+    // For just using signed transactions, it can just well be:
+    // MyRuntime: my_runtime::{ Module, Call, Storage, Event<T> }
+  }
+);
+```
+
+## Insert Keys in `service.rs`
+
+We have specify a local keystore with `KeyTypeId` to store app-specific sub-keys that are accessible by off-chain workers for signing transactions. But currently there is no sub-keys inside. We can add it in one of the following two ways.
+
+### Development: Add the First User Key as App Subkey
+
+In development environment, you can add the first user key as the app sub-key. Update the `service.rs` as followed.
+
+**node/src/service.rs**
+
+```rust
+pub fn new_full<C: Send + Default + 'static>(config: Configuration<C, GenesisConfig>)
+  -> Result<impl AbstractService, ServiceError>
+{
+  // ...
+  // ...
+  // Add this line
+  let dev_seed = config.dev_key_seed.clone();
+
+  // ...
+  // ...
+  let service = builder.with_network_protocol(|_| Ok(NodeProtocol::new()))?
+    .with_finality_proof_provider(|client, backend|
+      Ok(Arc::new(GrandpaFinalityProofProvider::new(backend, client)) as _)
+    )?
+    .build()?;
+
+  // Add the following section
+  if let Some(seed) = dev_seed {
+    service
+      .keystore()
+      .write()
+      .insert_ephemeral_from_seed_by_type::<runtime::my_runtime::crypto::Pair>(
+        &seed,
+        runtime::my_runtime::KEY_TYPE,
+      )
+      .expect("Dev Seed should always succeed.");
+  }
+}
+```
+
+Then you would be able to sign transactions, and this is good for **DEVELOPMENT only**.
+
+### Add an App Subkey via CLI
+
+In a more realistic setting, after setting up your substrate node, you can add a new app subkey via
+the command line interface as followed.
+
+```bash
+# Generate a new account
+$ subkey -s generate
+
+# Submit a new key via RPC
+$ http localhost:9933 jsonrpc=2.0 id=1 method=author_insertKey params:='["<YourKeyTypeId>", "<YourSeedPhrase>", "YourPublicKey"]'
```

Comment:
My debian system does not come with `http` installed. Maybe `curl` is more popular? Or this can be done through Apps UI. I wonder if we should have a separate doc on making RPC calls?

This comment is not a blocker. I'm just thinking out loud and I think competent developers will be able to take what you've written and figure out how to make the RPC call.

Reply:
Doing a simple doc/guide on RPC call will be great.

Just realized using `curl` is quite cumbersome, see the update. But this should be more widely-runnable to readers.